<a href="https://colab.research.google.com/github/Czrvillareal/activities/blob/main/lab1_titanic_villareal.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Run First


In [5]:
import pandas as pd
import matplotlib.pyplot as plt

df = pd.read_csv("https://raw.githubusercontent.com/Czrvillareal/activities/refs/heads/main/datasets/titanic.csv")

# Lab Instructions

### Step 1: Decomposition
        1. Open a Jupyter Notebook or any Python environment.
        2. Load the Titanic dataset and inspect its structure.

In [3]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [ ]:
rows, columns = df.shape
print(f"Number of rows: {rows}, Number of columns: {columns}")

Number of rows: 891, Number of columns: 12


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


## Question 1

Up to this point, determine the underlying tasks needed to be done for loading and handling the data based on
the steps described in this part? Write this as a text in your Colab notebook

  0. Programming Language, and the Notebook itself doesn't need to be mentioned however, they can still be considered as one of the tasks, as loading and handling wouldn't be possible if these weren't chosen beforehand. Hence, starting at step 0.
  1. **Preparing or determining the dataset**, in this case, titanic.csv.
  2. **Importing the libraries**, knowing the right libraries needed for this task, which was, pandas.
  3. **Connecting or making the dataset accessible**. There are several ways of retrieving the data source, in this case, titanic.csv was uploaded directly on the Github.
  4. **Using the appropriate code for inspecting**.
      - df.head() - was used to see a glimpse of what the table contains.
      - df.shape() - to identify the number of rows and columns without manually counting.
      - df.info() - to identify the general information about the table.


### Step 2: Pattern Recognition
        1. Now, lets analyze patterns in the dataset.
        2. Identify Missing Data. Run the following code "df.isnull().sum()" (Question 2)
        3. Explore Survival Rates. Run the following code "df['Survived'].mean()" (Question 3)

In [ ]:
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [42]:
rows = len(df['PassengerId'])
print(f"Number of rows: {rows}")

Number of rows: 891


In [41]:
df[df['Age'].isnull()].groupby(['Sex', 'Survived']).size()

Sex     Survived
female  0            17
        1            36
male    0           108
        1            16
dtype: int64

In [43]:
df[df['Cabin'].isnull()].groupby(['Sex', 'Survived']).size()

Sex     Survived
female  0            75
        1           142
male    0           406
        1            64
dtype: int64

In [44]:
df[df['Embarked'].isnull()].groupby(['Sex', 'Survived']).size()

Sex     Survived
female  1           2
dtype: int64

## Question 2

Using computation thinking, determine the blank values in the data. Write the algorithm as a text in your Google Colab

    - Which columns contain missing values? Should we remove or fill them?
      Answer:
        - Columns: Age(Fill), Cabin(Remove), Embarked(Fill/Remove) contain missing values.

1. Decomposition
    - Know the amount of rows (891)
    - Know the amount of columns (12)
    - Know if there are nulls (There are.)
    - How to handle the nulls?
2. Pattern Recognition
    - Three columns had nulls namely: Age, Cabin, Embarked
    - Missing values are as follows: Age(177), Cabin(687), Embarked(2)
    - Most of the null values come from Males who didnt survive, with the exemption of Embarked.
3. Abstraction
    - Ignore details regarding non-null values.
    - Focus on Column: Age, Cabin, Embarked.
4. Algorithm
    1. Load dataset, titanic.csv
    2. Check for missing values using df.info()
    3. Count missing values using df.isnull().sum()
    4. Handle missing values depending on how significant the data is:
        - **Cabin with 687 missing values**, despite being huge, the value isn't an integer and cannot be filled by any means, **can be removed**.
        - **Age with 177 missing values** will severely impact the analysis, the removal of this data can introduce bias, **therefore should be filled**.
        - **Embarked with 2 missing values** is little, **can be filled or removed.**

In [67]:
titanic_rate = df["Survived"].mean()

print(f"Survival Rate: {titanic_rate * 100: .2f}%")

Survival Rate:  38.38%


In [48]:
df.groupby(['Survived','Sex']).size()

Survived  Sex   
0         female     81
          male      468
1         female    233
          male      109
dtype: int64

In [66]:
male_rate = df[df["Sex"]== 'male']['Survived'].mean()
female_rate = df[df["Sex"]== 'female']['Survived'].mean()

print(f"Female survival rate: {female_rate * 100: .2f}%")
print(f"Male survival rate: {male_rate * 100: .2f}%")

Female survival rate:  74.20%
Male survival rate:  18.89%


In [54]:
df.groupby(['Survived','Pclass']).size()

Survived  Pclass
0         1          80
          2          97
          3         372
1         1         136
          2          87
          3         119
dtype: int64

In [69]:
c1_survivalRate = df[df["Pclass"]== 1]['Survived'].mean()
c2_survivalRate = df[df["Pclass"]== 2]['Survived'].mean()
c3_survivalRate = df[df["Pclass"]== 3]['Survived'].mean()

print(f"First Class: {c1_survivalRate * 100: .2f}%")
print(f"Second Class: {c2_survivalRate * 100: .2f}%")
print(f"Third Class: {c3_survivalRate * 100: .2f}%")

First Class:  62.96%
Second Class:  47.28%
Third Class:  24.24%


## Question 3

Using computation thinking, determine the survival rate of all passengers in the dataset. Write the algorithm as a text in your Google Colab.

      - What can you deduce on the survival rate of the passengers?
      Answer:
        - The survival rate in titanic is low with 38.38% meaning not even half of its passengers survive. Obviously, more female survived than male. Those in First Class are more likely to survive compared to Second and Third.

1. Decomposition
    - Are there Missing Values in Survived? (None)
    - What is the Survival Rate? (≈ 0.384)
    - What could be the possible factors? (Sex, Pclass)
2. Pattern Recognition
    - Class which are more likely to survive (Richer people, First Class)
    - Sex which are more likely to survive (Female)
3. Abstraction
    - Focus on Survived Column
4. Algorithm
    - Load dataset, titanic.csv
    - Compute total survivors and passengers
    - Calculate survival rate = total survivors/total passengers(alternatively, df["Survived"].mean())
    - Identify patterns(Which Class/Sex had more survivors)
    - Display Survival Rate for Pclass and Sex.
